### PR TITLE
[WIP] first draft of a BinaryProvider

### PR DESCRIPTION
**Pull request by staticfloat**
Mostly doing this to kick in CI

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,3 +10,13 @@
 # Work in progress
 
 Sketching out a reliable binary provider!
+
+# `@staticfloat`'s initial draft
+
+Borrowing inspiration from [Homebrew](https://github.com/Homebrew/brew/), this
+initial draft unpacks archives into separate folders within a `staging` area,
+then symlinks the resulting files into a `prefix`.
```
Comment (tkelman) [inline README.md]: Uh, why? Symlinks aren't going to work for libraries on Windows.

File: src/API.jl [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+"""
+`install(pkg::String, url::String; prefix::Prefix, verbose::Bool)`
+
+Given a `pkg` name, install it from `url` (either an HTTP/HTTPS url that will be
+downloaded from the internet, or a direct path to a local file) which should
+point to some kind of archive, e.g. a `.tar.gz` file, or a `.7z` file, etc...
+Install this `pkg` into the given `prefix` (which defualts to `global_prefix`),
+downloading, unpacking and linking it all in one fell swoop.
+"""
+function install(pkg::AbstractString, url::AbstractString;
```
Comment (tkelman) [inline src/API.jl]: we should require that all downloads are checksummed

File: README.md (hunk 1 above)
Comment (staticfloat) [inline README.md]: To try and solve the uninstallation problem; I've switched over to trying to determine installation receipts by asking our unpacking tools to give me a file manifest instead.  I didn't realize that Windows doesn't let you make symlinks unless you're an Administrator.  🦆

**Comment by staticfloat** [comment]
At this point I just need to sit down at a Windows machine and work out a few kinks.  I've backed off on the symlinks thanks to Windows.  Once I get Windows stuff working, I will sketch out a BinDeps provider so that this solution is backwards-compatible, even though I would like to replace BinDeps completely with this.

**Comment by staticfloat** [comment]
Alright, I think I have all the big pieces of machinery in place and adequately tested.  Now we just need to figure out the right API to expose to the user.  I have hacked together a BinDeps compatibility shim; you can use something similar to:
```julia
using BinDeps
using BinaryProvider

@BinDeps.setup

libtest = library_dependency("libtest")
@BP_provides("libtest", "http://foo.com/libtest.tar.gz", "a1b2c3....", libtest)

@BinDeps.install Dict(:libtest => :bindeps_libtest)
```

And it will install the requisite tarball into a project-specific prefix, then write out a `deps.jl` file with `bindeps_libtest` pointing to the installed library.


I'm not a huge fan of BinDeps though, and I'd like to make this setup easier to use by not requiring users to deal with BinDeps.  Although it's nice to have the option, I think ideally we don't need to deal with the BinDeps legacy.  So the alternative is to just do something like run `install()` directly within `build.jl` and have a helper function to emit something like `deps.jl` ourselves.  In this case, `deps.jl` would be as simple as running `using BinaryProvider; prefix = <path to prefix>; activate(prefix)` to setup the paths such that everything "just works" for the client package.

What are your thoughts?

**Comment by tkelman** [comment]
Why is it that you want to shove everything in a single prefix? That isn't going to play well with having multiple versions of the same package installed side by side. ccall does fine with absolute paths. Having one answer to where things are for linking them is a mostly build time problem.

**Comment by staticfloat** [comment]
> Why is it that you want to shove everything in a single prefix?

While things do default to a `global_prefix` if none is given, my intent is for each Julia package to have its own prefix.

**Comment by tkelman** [comment]
Let's not perpetuate bindeps mistake of using macros. Regular functions are much easier to follow and debug.

**Comment by tkelman** [comment]
If prefix management is optional, it's a lot of code complexity. I'd rather have this package download things and *nothing else*

**Comment by tkelman** [comment]
Even flexibility for different archive types I think is code we don't need - we can be as rigid as possible here to make this piece dead simple. So few upstream libraries provide binaries of any form, I think when that happens we can have maintainer-targeted scripts elsewhere to automate repackaging of existing binaries into the rigid single supported format we want. When things get compiled from source by the julia package author, we'll write the scaffolding that always packages things in our one format.

**Comment by staticfloat** [comment]
> Let's not perpetuate bindeps mistake of using macros.

The only reason I have a macro anywhere is because I want to automatically get the directory of the package invoking me.  If there's a non-macro way of doing that, I'd be happy to use it, otherwise the only macro being used is pretty short, only four lines long, and it's hardly what I would call complex.

> If prefix management is optional, it's a lot of code complexity.   I'd rather have this package download things and nothing else

I don't think we can do away with prefix management; so far this package is doing a bare minimum; it downloads things to a certain folder, and the whole idea of a "prefix" is basically a path with a special function defined to setup `PATH` and `DL_LOAD_PATH` such that you can find stuff within it.

**Comment by staticfloat** [comment]
> Even flexibility for different archive types I think is code we don't need

This I can agree with.  I'd be happy to choose one "blessed" format such as `.tar.gz` or something.

**Comment by tkelman** [comment]
Why would we need to touch PATH or DL_LOAD_PATH if we can know absolute paths of everything that's been downloaded?

**Comment by staticfloat** [comment]
As I see it, we have three options:

* Have this package download files and unpack them to a directory, and that's it.  Finding the interesting files (libraries, executables, etc...) is up to the client package, including complexity such as looking in `bin` for libraries on windows.  (We can't just put libraries in `lib` on windows because any included binaries will not be able to load them then).

* Have this package download files and unpack them to a directory, then find certain files using our own path searching/path resolution logic, perhaps as simple as just searching the appropriate directories (`bin` on Windows, `lib` elsewhere) for `$name.$(Libdl.dlext)` and mapping that to a variable that is defined within `deps.jl`.

* Have this package download files and unpack them to a directory, setup the proper paths at build time, then use the builtin search functionality (`ENV["PATH"]` and/or `DL_LOAD_PATH`) to find the interesting files, and write the absolute paths out to `deps.jl`.

I think the last option is best, because it seems like that's the option where we have the least amount of code we're writing either in this package or in client packages that use `BinaryProvider`.  To be clear, I'm not advocating that we set paths at client package load time, I'm saying we use the paths during build time, and anytime a user directly uses `BinaryProviders`.

**Comment by tkelman** [comment]
What is there to search for? If you're downloading something, you probably know what it is you're hoping to get out of that download.

**Comment by staticfloat** [comment]
In the simple cases, I agree with you.  But again, things like versioned/nonversioned sonames, putting things in `bin` on windows, versus `lib` on everything else, etc. make it nice and much simpler for everyone to use the builtin search functionality that we've already got in Base.

**Comment by tkelman** [comment]
That search functionality is going to find system libs and put us in compatibility hell if we let it. Touching PATH as a mechanism for finding libraries is asking for trouble.

If the package author controls the building of a binary archive, then the map from platform, to url and checksum, and files needed isn't dynamic at all. It can all be hardcoded. The only thing we'd need to do here is turn downloaded relative paths into const absolute paths for ccall.

**Comment by staticfloat** [comment]
> Touching PATH as a mechanism for finding libraries is asking for trouble.

I think you meant `DL_LOAD_PATH` instead of just `PATH`, and I disagree with you; this is the _primary_ method of finding libraries, and I think you're underestimating the number of problems that get solved by having an intelligent agent abstract away platform differences.

In any case, I don't think we're going to convince eachother without further proof, and the difference between my approach and your approach isn't very big, code-wise, so I think we should start building the mythical "scaffolding" for building these binaries that we want to install, and get some case studies to bring to the table.

**Comment by tkelman** [comment]
I don't think this should be allowed to find system libraries at all by default, not unless you explicitly opt into that. bin vs lib and shared library extension are pretty trivial to handle. sonames will be known ahead of time. What problem need abstracting?

**Comment by tkelman** [comment]
If you already know exactly where everything is because you're downloading a known binary, there's nothing to find.

**Pull request closed without merging**